packages/jazz-tools/src/schema/coMap.ts

```diff
@@ -0,0 +1,364 @@
+          unique?: CoValueUniqueness["uniqueness"];
+        }
+      | Account
+      | Group,
+  ) {
+    const schema = new this();
+
+    const { owner, uniqueness } = parseCoValueCreateOptions(options);
+    return CoMapInstance.fromInit(schema, init, owner, uniqueness);
+  }
+
+  /**
+   * Declare a Record-like CoMap schema, by extending `CoMap.Record(...)` and passing the value schema using `co`. Keys are always `string`.
+   *
+   * @example
+   * ```ts
+   * import { co, CoMap } from "jazz-tools";
+   *
+   * class ColorToFruitMap extends CoMap.Record(
+   *  co.ref(Fruit)
+   * ) {}
+   *
+   * // assume we have map: ColorToFruitMap
+   * // and strawberry: Fruit
+   * map["red"] = strawberry;
+   * ```
+   *
+   * @category Declaration
+   */
+  static Record<Value>(value: IfCo<Value, Value>) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+    class RecordLikeCoMap extends CoMapSchema {
+      [ItemsSym] = value;
+    }
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+    interface RecordLikeCoMap extends Record<string, Value> {}
+
+    return RecordLikeCoMap;
+  }
+}
+
+export type CoKeys<Map extends object> = Exclude<
+  keyof Map & string,
+  keyof CoMapSchema
+>;
+
+type ForceRequiredRef<V> = V extends co<InstanceType<
+  CoMapSchemaClass<any>
+> | null>
+  ? NonNullable<V>
+  : V extends co<CoMapSchema | undefined>
+    ? CoMapInstance<NonNullable<UnCo<V>>> | null
+    : V;
+
+export type CoMapInit<Map extends object> = {
+  [Key in CoKeys<Map> as undefined extends Map[Key]
+    ? never
+    : IfCo<Map[Key], Key>]: ForceRequiredRef<Map[Key]>;
+} & {
+  [Key in CoKeys<Map> as IfCo<Map[Key], Key>]?: ForceRequiredRef<Map[Key]>;
+};
+
+export class CoMapInstance<M extends CoMapSchema> extends CoValuePrototype {
+  declare $raw: RawCoMap;
+  declare $schema: M;
+  declare $id: ID<M>;
+  declare $resolutionNode: CoValueResolutionNode<M> | undefined;
+  refs = new Map<string, CoMapInstance<any>>();
+  private $lastUpdateTx: number;
+
+  constructor(
+    schema: M,
+    raw: RawCoMap,
+    resolutionNode?: CoValueResolutionNode<M>,
+  ) {
+    super();
+    this.$schema = schema;
+    this.$raw = raw;
+    this.$lastUpdateTx = raw.totalProcessedTransactions;
+    this.$id = raw.id as unknown as ID<M>;
+    this.$resolutionNode = resolutionNode;
+  }
+
+  static fromInit<M extends CoMapSchema>(
+    schema: M,
+    init: Simplify<CoMapInit<M>> | undefined,
+    owner: Account | Group,
+    uniqueness?: CoValueUniqueness,
+  ) {
+    const { raw, refs } = createCoMapFromInit(init, owner, schema, uniqueness);
+
+    return CoMapInstance.fromRaw(schema, raw, refs);
+  }
+
+  static fromRaw<M extends CoMapSchema>(
+    schema: M,
+    raw: RawCoMap,
+    refs?: Map<string, CoMapInstance<any> | undefined>,
+    resolutionNode?: CoValueResolutionNode<M>,
+  ) {
+    const instance = Object.create(
+      new CoMapInstance(schema, raw, resolutionNode),
+    );
```

Comment:
The CoMap values are assigned to the instance and everything else is part of the prototype.

This gives makes the CoMapInstnace easy to debug (the console log shows only the properties) and works great with JSON.stringify (see tests below)